caffe2/operators/gather_op.cc

```diff
@@ -75,43 +75,66 @@ print("OUTPUT:\n", workspace.FetchBlob("OUTPUT"))
     .Output(0, "OUTPUT", "Output tensor of rank $q+(r-1)$")
-    .TensorInferenceFunction([](const OperatorDef& /* unused */,
+    .TensorInferenceFunction([](const OperatorDef& def,
                                 const vector<TensorShape>& in) {
+      ArgumentHelper helper(def);
+      const int axis = helper.GetSingleArgument<int>("axis", 0);
+      const auto& data_dims = GetDimsVector(in[0]);
+      const auto& indices_dims = GetDimsVector(in[1]);
+
+      vector<int> output_dims =
+          caffe2::gather_helper::calc_output_shape_vector<int>(
+              data_dims, indices_dims, axis);
       vector<TensorShape> out(1);
-      for (auto d : in[1].dims()) {
-        out[0].add_dims(d);
-      }
-      for (int i = 1; i < in[0].dims_size(); ++i) {
-        out[0].add_dims(in[0].dims(i));
-      }
-      out[0].set_data_type(in[0].data_type());
+      out[0] = CreateTensorShape(output_dims, in[0].data_type());
       return out;
     });
 
 class GetGatherGradient : public GradientMakerBase {
   using GradientMakerBase::GradientMakerBase;
+
   vector<OperatorDef> GetGradientDefs() override {
     ArgumentHelper argsHelper(def_);
     const bool dense_gradient =
         argsHelper.GetSingleArgument<bool>("dense_gradient", false);
+    const int axis = argsHelper.GetSingleArgument<int>("axis", 0);
 
     using Op = GatherOp<CPUContext>;
 
-    if (dense_gradient) {
-      return vector<OperatorDef>{CreateOperatorDef(
-          "SparseToDense",
-          "",
-          vector<string>{I(Op::INDICES), GO(0), I(Op::DATA)},
-          vector<string>{GI(Op::DATA)})};
-    } else {
-      // For now we don't do any reshaping as the consumer of this op would
-      // probably be ScatterUpdate which is intenionally ignores shapes. We
-      // might need to revisit it in the future for correctness purposes. The
-      // right shape for the output woild be to flatten INDICES and collapse
-      // first X dims of GRAD
-      SetSparse(Op::DATA, I(Op::INDICES), GO(0));
-      return vector<OperatorDef>();
+    if (axis == 0) {
+      if (dense_gradient) {
+        return vector<OperatorDef>{CreateOperatorDef(
+            "SparseToDense",
+            "",
+            vector<string>{I(Op::INDICES), GO(0), I(Op::DATA)},
+            vector<string>{GI(Op::DATA)})};
+      } else {
+        // For now we don't do any reshaping as the consumer of this op would
+        // probably be ScatterUpdate which is intenionally ignores shapes. We
+        // might need to revisit it in the future for correctness purposes. The
+        // right shape for the output woild be to flatten INDICES and collapse
+        // first X dims of GRAD
+        SetSparse(Op::DATA, I(Op::INDICES), GO(0));
+        return vector<OperatorDef>();
+      }
     }
+
+    // TBD: This is misleading to use dense_gradient by default for axis 0
+    // and not othewise....
+    if (argsHelper.HasArgument("dense_gradient")) {
```

Comment:
I don't think we should enforce dense_gradient flag to be present if axis > 0. Effectively one can interpret `dense_gradient` as `force_dense_gradient` whereas def behavior is 'sparse only iff axis==0'

Reply:
This doesn't enforce dense_gradient, just checks its consistency for axis > 0 because there is no sparse impl for it  (unless you don't think that's good?). I'll rephrase the message.